### ISSUE TITLE
imread Should Raise If Path Doesn't Exist

### ISSUE THREAD

**Issue by zhanwenchen** [opencv/opencv#14095]
##### System information (version)
- OpenCV => 4.0.1
- Operating System / Platform => MacOS 10.14 Mojave
- Compiler => Xcode 10.1

##### Detailed description

`cv2.imread` doesn't check file existence and reads nonexistent files into a either a None object or an empty matrix. In comparison, `matplotlib.pyplot.imread` does throw; so does Pillow.

##### Steps to reproduce

```python
import cv2

empty_matrix = cv2.imread('nonexistent_path.jpg') # this is somehow legal
cv2.imshow(empty_matrix) # this will throw a "unable to convert NoneType to double" error
```

**Comment by alalek** [comment]
Mentioned API assumes that empty (`None` in Python) result is returned.
Changing this convention will break existed OpenCV applications.

**Comment by LaurentBerger** [comment]
@alalek What's about to add : 
`static const size_t CV_IO_IMREAD_EXCEPTION  = utils::getConfigurationParameterSizeT("OPENCVIO_IMREAD_EXCEPTION", 0);`

and new user can set to 1 if they want exception ?

**Comment by LJXLJXLJX** [comment]
Yes, without raising exceptions here, bugs can't be located in time.  
It may take a lot of time to debug manually to detect this tiny mistake.
I think this design does not meet the user-friendly concept.

**Comment by zhanwenchen** [comment]
@alalek I much prefer breaking changes to mismatched expectations. Why would this "break existed OpenCV applications?" Also why are those applications using this special fake "NoneType" thing anyway?

**Comment by mpolicki** [comment]
[Current imread documentation](https://docs.opencv.org/master/d4/da8/group__imgcodecs.html#ga288b8b3da0892bd651fce07b3bbd3a56) says: 

> If the image cannot be read (because of missing file, improper permissions, unsupported or invalid format), the function returns an empty matrix

From a software-development-best-practices standpoint, this is completely unacceptable. Nowhere will you find advice like the following:

> If an error occurs, don't report it to the caller, just fail silently.

Instead, the advice you typically find is [something like](https://opensource.com/article/17/5/30-best-practices-software-development-and-testing) (emphasis added):

> Check input and fail on nonsensical input or invalid state as early as possible, _preferably with an exception or error response that will make the exact problem clear to your caller_.

So, at the very least, `imread` needs to report to the caller which error it has encountered, if any. The mechanism of this error reporting should be appropriate to the language. In C++, that mechanism is exceptions. Therefore, `imread` should throw the appropriate exception if it encounters an error.

**Comment by sturkmen72** [comment]
i am working on https://github.com/opencv/opencv/pull/17753 and planning to work also on error handling. now waiting a review to continue

**Comment by alalek** [comment]
`imread()` comes from legacy C API which don't have C++ exception or similar error handling.
Adding exceptions into `imread()` directly break many applications.
So API itself should be evolved.
